core/trino-main/src/main/java/io/trino/sql/planner/JsonPredicateEvaluator.java

```diff
@@ -0,0 +1,512 @@
+            boolean leftHasScalar = false;
+            boolean leftHasNonScalar = false;
+            for (Object object : leftSequence) {
+                if (object instanceof JsonNode) {
+                    if (object instanceof NullNode) {
+                        leftHasJsonNull = true;
+                    }
+                    else if (((JsonNode) object).isValueNode()) {
+                        leftHasScalar = true;
+                    }
+                    else {
+                        leftHasNonScalar = true;
+                    }
+                }
+                else {
+                    leftHasScalar = true;
+                }
+            }
+
+            boolean rightHasJsonNull = false;
+            boolean rightHasScalar = false;
+            boolean rightHasNonScalar = false;
+            for (Object object : rightSequence) {
+                if (object instanceof JsonNode) {
+                    if (((JsonNode) object).isNull()) {
+                        rightHasJsonNull = true;
+                    }
+                    else if (((JsonNode) object).isValueNode()) {
+                        rightHasScalar = true;
+                    }
+                    else {
+                        rightHasNonScalar = true;
+                    }
+                }
+                else {
+                    rightHasScalar = true;
+                }
+            }
+
+            // try to find a quick error, i.e. a pair of left and right items which are of non-comparable categories
+            if (leftHasNonScalar && rightHasNonScalar ||
+                    leftHasNonScalar && rightHasScalar ||
+                    leftHasScalar && rightHasNonScalar) {
+                return null;
+            }
+
+            boolean found = false;
+
+            // try to find a quick null-based answer for == and <> operators
+            if (node.getOperator() == EQUAL && leftHasJsonNull && rightHasJsonNull) {
+                found = true;
+            }
+            if (node.getOperator() == NOT_EQUAL) {
+                if (leftHasJsonNull && (rightHasScalar || rightHasNonScalar) ||
+                        rightHasJsonNull && (leftHasScalar || leftHasNonScalar)) {
+                    found = true;
+                }
+            }
+            if (found && lax) {
+                return TRUE;
+            }
+
+            // compare scalars from left and right sequence
+            if (!leftHasScalar || !rightHasScalar) {
+                return found;
+            }
+            List<TypedValue> leftScalars = getScalars(leftSequence);
+            if (leftScalars == null) {
+                return null;
+            }
+            List<TypedValue> rightScalars = getScalars(rightSequence);
+            if (rightScalars == null) {
+                return null;
+            }
+            for (TypedValue leftValue : leftScalars) {
+                for (TypedValue rightValue : rightScalars) {
```

Comment:
as an improvement you may try to make it non quadratic for common cases (like equality comparison). Leave for now.